src/app/test_executive/slot_end_test.ml

```diff
@@ -0,0 +1,210 @@
+              { Test_Account.account_name =
+                  sprintf "%s-%d" sender_account_prefix i
+              ; balance = "1000"
+              ; timing = Untimed
+              } )
+    ; block_producers =
+        [ { node_name = "receiver"; account_name = "receiver-key" }
+        ; { node_name = "sender-1"; account_name = "sender-1-key" }
+        ; { node_name = "sender-2"; account_name = "sender-2-key" }
+        ; { node_name = "sender-3"; account_name = "sender-3-key" }
+        ]
+    ; snark_coordinator =
+        Some
+          { node_name = "snark-node"
+          ; account_name = "snark-node-key"
+          ; worker_nodes = 4
+          }
+    ; txpool_max_size = 10_000_000
+    ; snark_worker_fee = "0.0002"
+    ; num_archive_nodes = 0
+    ; proof_config =
+        { proof_config_default with
+          work_delay = Some 1
+        ; transaction_capacity =
+            Some Runtime_config.Proof_keys.Transaction_capacity.small
+        }
+    }
+
+  let fee = Currency.Fee.of_int 10_000_000
+
+  let amount = Currency.Amount.of_int 10_000_000
+
+  let tx_delay_ms = 5000
+
+  let run network t =
+    let open Malleable_error.Let_syntax in
+    let logger = Logger.create () in
+    let num_slots =
+      match (slot_tx_end, slot_chain_end) with
+      | None, None ->
+          10
+      | Some slot, None | None, Some slot | Some _, Some slot ->
+          Mina_numbers.Global_slot.to_int slot + 5
+    in
+    let receiver =
+      String.Map.find_exn (Network.block_producers network) "receiver"
+    in
+    let%bind receiver_pub_key = pub_key_of_node receiver in
+    let bp_senders =
+      String.Map.remove (Network.block_producers network) "receiver"
+      |> String.Map.data
+    in
+    let sender_kps =
+      String.Map.fold (Network.genesis_keypairs network) ~init:[]
+        ~f:(fun ~key ~data acc ->
+          if String.is_prefix key ~prefix:sender_account_prefix then data :: acc
+          else acc )
+    in
+    let sender_priv_keys =
+      List.map sender_kps ~f:(fun kp -> kp.keypair.private_key)
+    in
+    let pk_to_string = Signature_lib.Public_key.Compressed.to_base58_check in
+    [%log info] "receiver: %s" (pk_to_string receiver_pub_key) ;
+    let%bind () =
+      Malleable_error.List.iter sender_kps ~f:(fun s ->
+          let pk = s.keypair.public_key |> Signature_lib.Public_key.compress in
+          return ([%log info] "sender: %s" (pk_to_string pk)) )
+    in
+    let window_ms =
+      (Network.constraint_constants network).block_window_duration_ms
+    in
+    let all_nodes = Network.all_nodes network in
+    let%bind () =
+      wait_for t
+        (Wait_condition.nodes_to_initialize (String.Map.data all_nodes))
+    in
+    let%bind () =
+      section_hard "wait for 3 blocks to be produced (warm-up)"
+        (wait_for t (Wait_condition.blocks_to_be_produced 3))
+    in
+    let end_t =
+      Time.add (Time.now ())
+        (Time.Span.of_ms @@ float_of_int (num_slots * window_ms))
```

Comment:
This isn't necessarily aligned to the network's notion of slots. You probably want to compute an offset from the genesis timestamp.